### PR TITLE
Refactor resolver

### PR DESCRIPTION
**Pull request by spetrunin**
- Move processing of response data from datasource into resolver
- Add error locations
- Add errors when unable to resolve a field
- Add merging of upstream datasource errors

### CODE REVIEW

File: pkg/engine/datasource/graphql_datasource/graphql_datasource.go [hunk 1]
```diff
@@ -154,6 +150,10 @@ func (p *Planner) ConfigureFetch() plan.FetchConfiguration {
 		},
 		Variables:            p.variables,
 		DisallowSingleFlight: p.disallowSingleFlight,
+		ProcessResponseConfig: resolve.ProcessResponseConfig{
+			ExtractGraphqlResponse:    true,
```
Comment (jensneuse) [inline pkg/engine/datasource/graphql_datasource/graphql_datasource.go:154]: what happens if this is false? shouldn't it be the default?

File: pkg/engine/datasource/graphql_datasource/graphql_datasource.go [hunk 2]
```diff
@@ -869,24 +869,11 @@ func (f *Factory) Planner(ctx context.Context) plan.DataSourcePlanner {
...
-	}
-	entitiesPath = []string{"_entities", "[0]"}
-)
-
 type Source struct {
 	httpClient *http.Client
 }
 
-func (s *Source) Load(ctx context.Context, input []byte, bufPair *resolve.BufPair) (err error) {
+func (s *Source) Load(ctx context.Context, input []byte) (data []byte, err error) {
```
Comment (jensneuse) [inline pkg/engine/datasource/graphql_datasource/graphql_datasource.go:876]: wouldn't it better to use an io.Writer here?

File: pkg/engine/resolve/resolve.go [hunk 3]
```diff
@@ -237,7 +250,7 @@ type Fetch interface {
 type Fetches []Fetch
 
 type DataSource interface {
-	Load(ctx context.Context, input []byte, bufPair *BufPair) (err error)
+	Load(ctx context.Context, input []byte) (data []byte, err error)
```
Comment (jensneuse) [inline pkg/engine/resolve/resolve.go:253]: I don't like returning byte slice, I'd much rather use io.Writer as argument and return err only

File: pkg/engine/datasource/graphql_datasource/graphql_datasource.go (hunk 1 above)
Comment (spetrunin) [reply pkg/engine/datasource/graphql_datasource/graphql_datasource.go:154]: it is always false for all other datasources

I was choosing between 2 approaches `NonGraphqlResponse` and `GraphqlResponse`

so it is like set to `false` for all other and future data sources or set `true` for the graphql data source

Comment (jensneuse) [reply pkg/engine/datasource/graphql_datasource/graphql_datasource.go:154]: what's the expectation on "non graphql" datasources? What format should they return?
wouldn't it be easier if they just return {"errors":null,"data":null} format as well and we treat it like a standard?

**Review by jensneuse — APPROVED** [review]

**Review by pvormste — APPROVED** [review]

**Review by jensneuse — APPROVED** [review]

**Pull request merged**